<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kaixuan Luo" />

<meta name="date" content="2022-05-25" />

<title>SCEPTRE analysis on LUHMES CROP-seq data</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA_analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">SCEPTRE analysis on LUHMES CROP-seq
data</h1>
<h4 class="author">Kaixuan Luo</h4>
<h4 class="date">2022-05-25</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-09-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSFA_analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220524code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220524)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220524code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220524)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomkevinlkxGSFAanalysistreee5f4b4a0565745cd89f31f1ec64d83ffcfb1d24etargetblanke5f4b4aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/kevinlkx/GSFA-analysis/tree/e5f4b4a0565745cd89f31f1ec64d83ffcfb1d24e" target="_blank">e5f4b4a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomkevinlkxGSFAanalysistreee5f4b4a0565745cd89f31f1ec64d83ffcfb1d24etargetblanke5f4b4aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/kevinlkx/GSFA-analysis/tree/e5f4b4a0565745cd89f31f1ec64d83ffcfb1d24e" target="_blank">e5f4b4a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  Rplots.pdf
    Untracked:  analysis/check_Tcells_datasets.Rmd
    Untracked:  analysis/fscLVM_analysis.Rmd
    Untracked:  analysis/spca_LUHMES_data.Rmd
    Untracked:  analysis/test_seurat.Rmd
    Untracked:  code/gsfa_negctrl_job.sbatch
    Untracked:  code/music_LUHMES_Yifan.R
    Untracked:  code/plotting_functions.R
    Untracked:  code/run_fscLVM_LUHMES_data.R
    Untracked:  code/run_gsfa_2groups_negctrl.R
    Untracked:  code/run_gsfa_negctrl.R
    Untracked:  code/run_music_LUHMES.R
    Untracked:  code/run_music_LUHMES_data.sbatch
    Untracked:  code/run_music_LUHMES_data_20topics.R
    Untracked:  code/run_music_LUHMES_data_20topics.sbatch
    Untracked:  code/run_sceptre_Tcells_data.sbatch
    Untracked:  code/run_sceptre_Tcells_stimulated_data.sbatch
    Untracked:  code/run_sceptre_Tcells_test_data.sbatch
    Untracked:  code/run_sceptre_Tcells_unstimulated_data.sbatch
    Untracked:  code/run_sceptre_permuted_data.sbatch
    Untracked:  code/run_spca_LUHMES.R
    Untracked:  code/run_spca_TCells.R
    Untracked:  code/run_twostep_clustering_LUHMES_data.sbatch
    Untracked:  code/run_twostep_clustering_Tcells_data.sbatch
    Untracked:  code/run_unguided_gsfa_LUHMES.R
    Untracked:  code/run_unguided_gsfa_LUHMES.sbatch
    Untracked:  code/run_unguided_gsfa_Tcells.R
    Untracked:  code/run_unguided_gsfa_Tcells.sbatch
    Untracked:  code/sceptre_LUHMES_data.R
    Untracked:  code/sceptre_Tcells_stimulated_data.R
    Untracked:  code/sceptre_Tcells_unstimulated_data.R
    Untracked:  code/sceptre_permutation_analysis.R
    Untracked:  code/sceptre_permute_analysis.R
    Untracked:  code/seurat_sim_fpr_tpr.R
    Untracked:  code/unguided_GFSA_mixture_normal_prior.cpp

Unstaged changes:
    Modified:   analysis/index.Rmd
    Modified:   analysis/music_LUHMES_data.Rmd
    Modified:   analysis/sceptre_TCells_data.Rmd
    Modified:   analysis/twostep_clustering_LUHMES_data.Rmd
    Modified:   code/run_sceptre_LUHMES_data.R
    Modified:   code/run_sceptre_LUHMES_data.sbatch
    Modified:   code/run_sceptre_LUHMES_permuted_data.R
    Modified:   code/run_sceptre_Tcells_permuted_data.R
    Modified:   code/run_sceptre_cropseq_data.sbatch
    Modified:   code/run_twostep_clustering_LUHMES_data.R
    Modified:   code/sceptre_analysis.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/sceptre_LUHMES_data.Rmd</code>)
and HTML (<code>docs/sceptre_LUHMES_data.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/e5f4b4a0565745cd89f31f1ec64d83ffcfb1d24e/analysis/sceptre_LUHMES_data.Rmd" target="_blank">e5f4b4a</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-09-16
</td>
<td>
updated results with new version and added permutation result
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/0b325fb6090b54ffa296674f4097df62bb5fc34f/docs/sceptre_LUHMES_data.html" target="_blank">0b325fb</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/4b70cfd55bbfe76abaff21652a4d7fb5fc6e14e0/analysis/sceptre_LUHMES_data.Rmd" target="_blank">4b70cfd</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
updated code links
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/a38de80af621969cd22f8872c5be68effa614f35/docs/sceptre_LUHMES_data.html" target="_blank">a38de80</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/144346c186c1d61ebc7e712be653539c467a224a/analysis/sceptre_LUHMES_data.Rmd" target="_blank">144346c</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
updated document format
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/dab64ad6af4892b649e2c4b4e5d36b3d23346298/docs/sceptre_LUHMES_data.html" target="_blank">dab64ad</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/c21a39fc5e7446a1e4fde854489897d9598ca09e/analysis/sceptre_LUHMES_data.Rmd" target="_blank">c21a39f</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-05-26
</td>
<td>
SCEPTRE analysis on LUHMES data
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="about-the-data-sets" class="section level2">
<h2>About the data sets</h2>
<p>CROP-seq datasets:
<code>/project2/xinhe/yifan/Factor_analysis/shared_data/</code> The data
are Seurat objects, with raw gene counts stored in <a
href="mailto:obj@assays" class="email">obj@assays</a><span
class="math inline">\(RNA@counts, and cell meta data stored in
obj@meta.data. Normalized and scaled data used for GSFA are stored in
obj@assays\)</span><a href="mailto:RNA@scale.data"
class="email">RNA@scale.data</a> , the rownames of which are the 6k
genes used for GSFA.</p>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<p>Scripts for running SCEPTRE:</p>
<ul>
<li><a
href="https://github.com/kevinlkx/GSFA_analysis/tree/main/code/run_sceptre_LUHMES_data.R">R
script</a>.</li>
<li><a
href="https://github.com/kevinlkx/GSFA_analysis/tree/main/code/run_sceptre_LUHMES_data.sbatch">sbatch
script</a>.</li>
</ul>
<pre class="bash"><code>mkdir -p /project2/xinhe/kevinluo/GSFA/sceptre_analysis/log
cd /project2/xinhe/kevinluo/GSFA/sceptre_analysis/log

sbatch --mem=50G --cpus-per-task=10 --partition=xinhe ~/projects/GSFA_analysis/code/run_sceptre_LUHMES_data.sbatch</code></pre>
<p>Permutation analysis</p>
<pre class="bash"><code>mkdir -p /project2/xinhe/kevinluo/GSFA/sceptre_analysis/log
cd /project2/xinhe/kevinluo/GSFA/sceptre_analysis/log

for i in {1..10}
do
  sbatch --mem=50G --cpus-per-task=10 ~/projects/GSFA_analysis/code/run_sceptre_LUHMES_permuted_data.sbatch $i
done

# run_sceptre_LUHMES_permuted_data.sbatch runs run_sceptre_LUHMES_permuted_data.R</code></pre>
<p>Load packages</p>
<pre class="r"><code>dyn.load(&#39;/software/geos-3.7.0-el7-x86_64/lib64/libgeos_c.so&#39;) # attach the geos lib for Seurat
suppressPackageStartupMessages(library(tidyverse))
library(cowplot)
library(Matrix)
library(sceptre)
library(Seurat)
source(&quot;code/plotting_functions.R&quot;)</code></pre>
<div id="prepare-input-data-for-sceptre" class="section level3">
<h3>Prepare input data for SCEPTRE</h3>
<p>Load the Seurat object of LUHMES data</p>
<pre class="r"><code>LUHMES_data &lt;- readRDS(&#39;/project2/xinhe/yifan/Factor_analysis/shared_data/LUHMES_cropseq_data_seurat.rds&#39;)
datadir &lt;- &#39;/project2/xinhe/kevinluo/GSFA/sceptre_analysis/LUHMES_cropseq_data&#39;</code></pre>
<p>We first prepare three separate data objects required to run SCEPTRE:
the gene expression matrix, the perturbation matrix, and the covariate
matrix.</p>
<ol style="list-style-type: decimal">
<li>Gene expression matrices Gene expression (gene-by-cell) raw count
matrix</li>
</ol>
<pre class="r"><code>gene_matrix &lt;- LUHMES_data@assays$RNA@counts
# gene-by-cell expression matrix
gene_matrix[1:10, 1:3]
dim(gene_matrix)
# 10 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
#                 Run1_AAACCTGAGATCGATA Run1_AAACCTGCAGCTCGAC
# ENSG00000243485                     .                     .
# ENSG00000237613                     .                     .
# ENSG00000186092                     .                     .
# ENSG00000238009                     .                     .
# ENSG00000239945                     .                     .
# ENSG00000239906                     .                     .
# ENSG00000241599                     .                     .
# ENSG00000279928                     .                     .
# ENSG00000279457                     1                     .
# ENSG00000228463                     1                     .
#                 Run1_AAACCTGCATTAGCCA
# ENSG00000243485                     .
# ENSG00000237613                     .
# ENSG00000186092                     .
# ENSG00000238009                     .
# ENSG00000239945                     .
# ENSG00000239906                     .
# ENSG00000241599                     .
# ENSG00000279928                     .
# ENSG00000279457                     1
# ENSG00000228463                     .
# [1] 33694  8708</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Cell meta data and covariate matrix</li>
</ol>
<pre class="r"><code>metadata &lt;- LUHMES_data@meta.data
metadata[1:5, ]
covariate_matrix &lt;- metadata[,c(&#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;percent_mt&#39;)]
covariate_matrix[1:5,]
dim(covariate_matrix)
#                        orig.ident nCount_RNA nFeature_RNA ADNP ARID1B ASH1L
# Run1_AAACCTGAGATCGATA LUHMES_Run1      10655         3287    0      0     0
# Run1_AAACCTGCAGCTCGAC LUHMES_Run1       8648         2791    0      0     0
# Run1_AAACCTGCATTAGCCA LUHMES_Run1       9015         2922    0      0     0
# Run1_AAACCTGGTAAGGGCT LUHMES_Run1       9090         2925    0      0     0
# Run1_AAACCTGGTTCCGTCT LUHMES_Run1      10182         3052    0      0     0
#                       CHD2 CHD8 CTNND2 DYRK1A HDAC5 MECP2 MYT1L Nontargeting
# Run1_AAACCTGAGATCGATA    0    1      0      0     0     0     0            0
# Run1_AAACCTGCAGCTCGAC    0    0      0      0     0     0     0            0
# Run1_AAACCTGCATTAGCCA    1    0      0      0     0     0     0            0
# Run1_AAACCTGGTAAGGGCT    0    0      0      1     0     0     0            0
# Run1_AAACCTGGTTCCGTCT    0    0      0      0     1     0     0            0
#                       POGZ PTEN RELN SETD5 percent_mt
# Run1_AAACCTGAGATCGATA    0    0    0     0   3.594557
# Run1_AAACCTGCAGCTCGAC    0    0    1     0   2.717391
# Run1_AAACCTGCATTAGCCA    0    0    0     0   2.995008
# Run1_AAACCTGGTAAGGGCT    0    0    0     0   3.080308
# Run1_AAACCTGGTTCCGTCT    0    0    0     0   3.142801
#                        orig.ident nCount_RNA nFeature_RNA percent_mt
# Run1_AAACCTGAGATCGATA LUHMES_Run1      10655         3287   3.594557
# Run1_AAACCTGCAGCTCGAC LUHMES_Run1       8648         2791   2.717391
# Run1_AAACCTGCATTAGCCA LUHMES_Run1       9015         2922   2.995008
# Run1_AAACCTGGTAAGGGCT LUHMES_Run1       9090         2925   3.080308
# Run1_AAACCTGGTTCCGTCT LUHMES_Run1      10182         3052   3.142801
# [1] 8708    4</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Perturbation matrix (a binary matrix of perturbations, rows are gRNA
groups and columns are cell barcodes)</li>
</ol>
<pre class="r"><code>combined_perturbation_matrix &lt;- t(metadata[,4:18])
dim(combined_perturbation_matrix)
combined_perturbation_matrix[1:10,1:3]
range(combined_perturbation_matrix)
# [1]   15 8708
#        Run1_AAACCTGAGATCGATA Run1_AAACCTGCAGCTCGAC Run1_AAACCTGCATTAGCCA
# ADNP                       0                     0                     0
# ARID1B                     0                     0                     0
# ASH1L                      0                     0                     0
# CHD2                       0                     0                     1
# CHD8                       1                     0                     0
# CTNND2                     0                     0                     0
# DYRK1A                     0                     0                     0
# HDAC5                      0                     0                     0
# MECP2                      0                     0                     0
# MYT1L                      0                     0                     0
# [1] 0 1</code></pre>
<ol start="4" style="list-style-type: decimal">
<li>Specify the gene-gRNA group pairs to test for association</li>
</ol>
<p>Include the 6k genes used for GSFA as candidates</p>
<pre class="r"><code># Normalized and scaled data used for GSFA, the rownames of which are the 6k genes used for GSFA
scaled_gene_matrix &lt;- LUHMES_data@assays$RNA@scale.data
dim(scaled_gene_matrix)
selected_gene_id &lt;- rownames(scaled_gene_matrix)
all(selected_gene_id %in% rownames(gene_matrix))
# [1] 6000 8708
# [1] TRUE</code></pre>
<pre class="r"><code>gRNA_group &lt;- rownames(combined_perturbation_matrix)
pairs &lt;- expand.grid(selected_gene_id, gRNA_group)
gene_gRNA_group_pairs &lt;- data.frame(gene_id = pairs$Var1, gRNA_group = pairs$Var2, pair_type = &quot;candidate&quot;)
gene_gRNA_group_pairs[gene_gRNA_group_pairs$gRNA_group == &quot;Nontargeting&quot;, &quot;pair_type&quot;] &lt;- &quot;negative_control&quot;
table(gene_gRNA_group_pairs$pair_type)
table(gene_gRNA_group_pairs$gRNA_group)
dim(gene_gRNA_group_pairs)
# 
#        candidate negative_control 
#            84000             6000 
# 
#         ADNP       ARID1B        ASH1L         CHD2         CHD8       CTNND2 
#         6000         6000         6000         6000         6000         6000 
#       DYRK1A        HDAC5        MECP2        MYT1L Nontargeting         POGZ 
#         6000         6000         6000         6000         6000         6000 
#         PTEN         RELN        SETD5 
#         6000         6000         6000 
# [1] 90000     3</code></pre>
<pre class="r"><code>save(list = c(&quot;gene_matrix&quot;, &quot;combined_perturbation_matrix&quot;, &quot;covariate_matrix&quot;),
     file = file.path(datadir, &#39;data.matrices.RData&#39;))
saveRDS(gene_gRNA_group_pairs, file.path(datadir, &quot;gene.gRNA.group.pairs.rds&quot;))</code></pre>
</div>
<div id="analyze-the-sceptre-results" class="section level3">
<h3>Analyze the SCEPTRE results</h3>
<pre class="r"><code>outdir &lt;- &#39;/project2/xinhe/kevinluo/GSFA/sceptre_analysis/LUHMES_data_updated/sceptre_output&#39;
result &lt;- readRDS(file.path(outdir, &#39;sceptre.result.rds&#39;))
head(result, 10)
#            gene_id gRNA_id pair_type      p_value     z_value log_fold_change
# 1  ENSG00000251562    ADNP candidate 8.276198e-04  -2.7128544    -0.042399626
# 2  ENSG00000167552    ADNP candidate 5.699734e-05   5.6529476     0.090952484
# 3  ENSG00000205542    ADNP candidate 1.175300e-03  -3.2572089    -0.077037279
# 4  ENSG00000104722    ADNP candidate 5.498633e-07 -11.9722455    -0.218557354
# 5  ENSG00000156508    ADNP candidate 6.863815e-01  -0.4263849    -0.040725865
# 6  ENSG00000034510    ADNP candidate 5.594500e-02   2.2087809     0.025101774
# 7  ENSG00000198712    ADNP candidate 8.261282e-01   0.1660938    -0.005275184
# 8  ENSG00000213190    ADNP candidate 1.769592e-01   1.4053725     0.012649649
# 9  ENSG00000184009    ADNP candidate 5.329976e-01   0.7832493    -0.007898580
# 10 ENSG00000198668    ADNP candidate 7.860573e-03   4.6174578     0.042193477</code></pre>
<p>Negative control pairs</p>
<pre class="r"><code>neg_control_p_vals &lt;- result %&gt;% filter(pair_type == &quot;negative_control&quot;) %&gt;% pull(p_value)
make_qq_plot(neg_control_p_vals)</code></pre>
<p><img src="figure/sceptre_LUHMES_data.Rmd/unnamed-chunk-5-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>Candidate pairs We extract the p-values corresponding to the
candidate pairs and apply a Benjamini-Hochberg (BH) correction to adjust
for multiple testing.</p>
<pre class="r"><code>candidate_pair_results &lt;- result %&gt;% filter(pair_type == &quot;candidate&quot;) %&gt;%
  mutate(p_val_adj = p.adjust(p_value, method = &quot;BH&quot;))
head(candidate_pair_results)

candidate_pvals &lt;- candidate_pair_results %&gt;% pull(p_value)
make_qq_plot(candidate_pvals)</code></pre>
<p><img src="figure/sceptre_LUHMES_data.Rmd/fdr-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre><code>#           gene_id gRNA_id pair_type      p_value     z_value log_fold_change
# 1 ENSG00000251562    ADNP candidate 8.276198e-04  -2.7128544     -0.04239963
# 2 ENSG00000167552    ADNP candidate 5.699734e-05   5.6529476      0.09095248
# 3 ENSG00000205542    ADNP candidate 1.175300e-03  -3.2572089     -0.07703728
# 4 ENSG00000104722    ADNP candidate 5.498633e-07 -11.9722455     -0.21855735
# 5 ENSG00000156508    ADNP candidate 6.863815e-01  -0.4263849     -0.04072587
# 6 ENSG00000034510    ADNP candidate 5.594500e-02   2.2087809      0.02510177
#      p_val_adj
# 1 0.1089391252
# 2 0.0160663656
# 3 0.1220336025
# 4 0.0003446905
# 5 0.9638421091
# 6 0.6058244221</code></pre>
<p>We call pairs with an adjusted p-value of less or equal than 0.1
significant; the discovery set (i.e., the set of significant pairs) has
a false discovery rate (FDR) of 10%.</p>
<pre class="r"><code>discovery_set &lt;- candidate_pair_results %&gt;% filter(p_val_adj &lt;= 0.1)
head(discovery_set)
#           gene_id gRNA_id pair_type      p_value    z_value log_fold_change
# 1 ENSG00000167552    ADNP candidate 5.699734e-05   5.652948      0.09095248
# 2 ENSG00000104722    ADNP candidate 5.498633e-07 -11.972246     -0.21855735
# 3 ENSG00000124766    ADNP candidate 2.428708e-06  -5.419528     -0.08567983
# 4 ENSG00000173267    ADNP candidate 2.220446e-16  12.720042      0.17604320
# 5 ENSG00000131711    ADNP candidate 1.852806e-05   4.984110      0.08523870
# 6 ENSG00000277586    ADNP candidate 6.380460e-05  -6.488863     -0.15174885
#      p_val_adj
# 1 1.606637e-02
# 2 3.446905e-04
# 3 1.214354e-03
# 4 1.434750e-12
# 5 6.539316e-03
# 6 1.763022e-02</code></pre>
</div>
<div id="permutation-analysis" class="section level3">
<h3>Permutation analysis</h3>
<p>Pool results from 10 permuted data sets</p>
<pre class="r"><code>outdir &lt;- &quot;/project2/xinhe/kevinluo/GSFA/sceptre_analysis/LUHMES_data_updated/sceptre_output/permutation&quot;

sceptre_res &lt;- data.frame()
for(permute_num in 1:10){
  cat(&quot;permutation set: &quot;,permute_num, &quot;\n&quot;)
  res_dir &lt;- paste0(outdir, &quot;/perm_&quot;, permute_num)
  res &lt;- readRDS(file.path(res_dir, &#39;sceptre.result.rds&#39;))
  sceptre_res &lt;- rbind(sceptre_res, data.frame(permute_num = permute_num, res))
}

nrow(sceptre_res)
anyNA(sceptre_res$p_value)
# permutation set:  1 
# permutation set:  2 
# permutation set:  3 
# permutation set:  4 
# permutation set:  5 
# permutation set:  6 
# permutation set:  7 
# permutation set:  8 
# permutation set:  9 
# permutation set:  10 
# [1] 900000
# [1] FALSE</code></pre>
<p>QQ-plot for candidate pairs</p>
<pre class="r"><code>candidate_pair_results &lt;- sceptre_res %&gt;% filter(pair_type == &quot;candidate&quot;) %&gt;%
  mutate(p_val_adj = p.adjust(p_value, method = &quot;BH&quot;))

candidate_p_vals &lt;- candidate_pair_results %&gt;% pull(p_value)</code></pre>
<pre class="r"><code># pdf(file.path(outdir, &quot;LUHMES_permutation_qqplot.pdf&quot;), width = 5, height = 5)
qqunif.plot(candidate_p_vals, main = &quot;SCEPTRE&quot;)
# Loading required package: grid</code></pre>
<p><img src="figure/sceptre_LUHMES_data.Rmd/qqplot_permutation-1.png" width="600" style="display: block; margin: auto;" /></p>
<pre class="r"><code># dev.off()</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.2.0 (2022-04-22)
# Platform: x86_64-pc-linux-gnu (64-bit)
# Running under: CentOS Linux 7 (Core)
# 
# Matrix products: default
# BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so
# 
# locale:
#  [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C           
#  [4] LC_COLLATE=C         LC_MONETARY=C        LC_MESSAGES=C       
#  [7] LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C        
# [10] LC_TELEPHONE=C       LC_MEASUREMENT=C     LC_IDENTIFICATION=C 
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
#  [1] lattice_0.20-45    sp_1.4-7           SeuratObject_4.1.0 Seurat_4.1.1      
#  [5] sceptre_0.1.0      Matrix_1.4-1       cowplot_1.1.1      forcats_0.5.1     
#  [9] stringr_1.4.0      dplyr_1.0.9        purrr_0.3.4        readr_2.1.2       
# [13] tidyr_1.2.0        tibble_3.1.7       ggplot2_3.3.6      tidyverse_1.3.1   
# [17] workflowr_1.7.0   
# 
# loaded via a namespace (and not attached):
#   [1] readxl_1.4.0          backports_1.4.1       plyr_1.8.7           
#   [4] igraph_1.3.4          lazyeval_0.2.2        splines_4.2.0        
#   [7] listenv_0.8.0         scattermore_0.8       digest_0.6.29        
#  [10] htmltools_0.5.2       fansi_1.0.3           magrittr_2.0.3       
#  [13] tensor_1.5            cluster_2.1.3         ROCR_1.0-11          
#  [16] tzdb_0.3.0            globals_0.15.0        modelr_0.1.8         
#  [19] matrixStats_0.62.0    spatstat.sparse_2.1-1 colorspace_2.0-3     
#  [22] rvest_1.0.2           ggrepel_0.9.1         haven_2.5.0          
#  [25] xfun_0.30             callr_3.7.0           crayon_1.5.1         
#  [28] jsonlite_1.8.0        progressr_0.10.0      spatstat.data_2.2-0  
#  [31] survival_3.3-1        zoo_1.8-10            glue_1.6.2           
#  [34] polyclip_1.10-0       gtable_0.3.0          leiden_0.4.2         
#  [37] future.apply_1.9.0    abind_1.4-5           scales_1.2.0         
#  [40] DBI_1.1.3             spatstat.random_2.2-0 miniUI_0.1.1.1       
#  [43] Rcpp_1.0.8.3          viridisLite_0.4.0     xtable_1.8-4         
#  [46] reticulate_1.24       spatstat.core_2.4-2   htmlwidgets_1.5.4    
#  [49] httr_1.4.3            RColorBrewer_1.1-3    ellipsis_0.3.2       
#  [52] ica_1.0-2             farver_2.1.0          pkgconfig_2.0.3      
#  [55] uwot_0.1.11           sass_0.4.1            dbplyr_2.1.1         
#  [58] deldir_1.0-6          utf8_1.2.2            tidyselect_1.1.2     
#  [61] rlang_1.0.2           reshape2_1.4.4        later_1.3.0          
#  [64] munsell_0.5.0         cellranger_1.1.0      tools_4.2.0          
#  [67] cli_3.3.0             generics_0.1.2        broom_0.8.0          
#  [70] ggridges_0.5.3        evaluate_0.15         fastmap_1.1.0        
#  [73] yaml_2.3.5            goftest_1.2-3         processx_3.5.3       
#  [76] knitr_1.39            fs_1.5.2              fitdistrplus_1.1-8   
#  [79] RANN_2.6.1            pbapply_1.5-0         future_1.25.0        
#  [82] nlme_3.1-157          whisker_0.4           mime_0.12            
#  [85] xml2_1.3.3            compiler_4.2.0        rstudioapi_0.13      
#  [88] plotly_4.10.0         png_0.1-7             spatstat.utils_2.3-1 
#  [91] reprex_2.0.1          bslib_0.3.1           stringi_1.7.6        
#  [94] highr_0.9             ps_1.7.0              rgeos_0.5-9          
#  [97] vctrs_0.4.1           pillar_1.7.0          lifecycle_1.0.1      
# [100] spatstat.geom_2.4-0   lmtest_0.9-40         jquerylib_0.1.4      
# [103] RcppAnnoy_0.0.19      data.table_1.14.2     irlba_2.3.5          
# [106] httpuv_1.6.5          patchwork_1.1.1       R6_2.5.1             
# [109] promises_1.2.0.1      KernSmooth_2.23-20    gridExtra_2.3        
# [112] parallelly_1.31.1     codetools_0.2-18      MASS_7.3-56          
# [115] assertthat_0.2.1      rprojroot_2.0.3       withr_2.5.0          
# [118] sctransform_0.3.3     mgcv_1.8-40           parallel_4.2.0       
# [121] hms_1.1.1             rpart_4.1.16          rmarkdown_2.14       
# [124] Rtsne_0.16            git2r_0.30.1          getPass_0.2-2        
# [127] shiny_1.7.1           lubridate_1.8.0</code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
