<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kaixuan Luo" />

<meta name="date" content="2022-08-10" />

<title>Two step clustering analysis on CD8+ T Cell (stimulated) CROP-seq data</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA_analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Two step clustering analysis on CD8+ T Cell (stimulated) CROP-seq data</h1>
<h4 class="author">Kaixuan Luo</h4>
<h4 class="date">2022-08-10</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-08-31
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSFA_analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220524code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220524)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220524code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220524)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomkevinlkxGSFAanalysistree2d2298aedcf31771ead039a689ce7d4e334b0928targetblank2d2298aa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/kevinlkx/GSFA-analysis/tree/2d2298aedcf31771ead039a689ce7d4e334b0928" target="_blank">2d2298a</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomkevinlkxGSFAanalysistree2d2298aedcf31771ead039a689ce7d4e334b0928targetblank2d2298aa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/kevinlkx/GSFA-analysis/tree/2d2298aedcf31771ead039a689ce7d4e334b0928" target="_blank">2d2298a</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  Rplots.pdf
    Untracked:  analysis/check_Tcells_datasets.Rmd
    Untracked:  analysis/interpret_gsfa_LUHMES.Rmd
    Untracked:  analysis/interpret_gsfa_TCells.Rmd
    Untracked:  analysis/spca_LUHMES_data.Rmd
    Untracked:  analysis/test_seurat.Rmd
    Untracked:  code/gsfa_negctrl_job.sbatch
    Untracked:  code/music_LUHMES_Yifan.R
    Untracked:  code/plotting_functions.R
    Untracked:  code/run_gsfa_2groups_negctrl.R
    Untracked:  code/run_gsfa_negctrl.R
    Untracked:  code/run_music_LUHMES.R
    Untracked:  code/run_music_LUHMES_data.sbatch
    Untracked:  code/run_sceptre_LUHMES_data.sbatch
    Untracked:  code/run_sceptre_Tcells_stimulated_data.sbatch
    Untracked:  code/run_sceptre_Tcells_unstimulated_data.sbatch
    Untracked:  code/run_spca_LUHMES.R
    Untracked:  code/run_spca_TCells.R
    Untracked:  code/run_unguided_gsfa_LUHMES.R
    Untracked:  code/run_unguided_gsfa_LUHMES.sbatch
    Untracked:  code/run_unguided_gsfa_Tcells.R
    Untracked:  code/run_unguided_gsfa_Tcells.sbatch
    Untracked:  code/sceptre_LUHMES_data.R
    Untracked:  code/sceptre_Tcells_stimulated_data.R
    Untracked:  code/sceptre_Tcells_unstimulated_data.R
    Untracked:  code/seurat_sim_fpr_tpr.R
    Untracked:  code/unguided_GFSA_mixture_normal_prior.cpp

Unstaged changes:
    Modified:   analysis/sceptre_LUHMES_data.Rmd
    Modified:   code/run_sceptre_cropseq_data.sbatch
    Modified:   code/sceptre_analysis.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/twostep_clustering_TCells_data.Rmd</code>) and HTML (<code>docs/twostep_clustering_TCells_data.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/2d2298aedcf31771ead039a689ce7d4e334b0928/analysis/twostep_clustering_TCells_data.Rmd" target="_blank">2d2298a</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-31
</td>
<td>
added qqplots for combined results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/104e3915d02deea37cfb2afda409088498058f32/docs/twostep_clustering_TCells_data.html" target="_blank">104e391</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-25
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/a27762bf831c5b690dc92314cf11ad36f1e1ae56/analysis/twostep_clustering_TCells_data.Rmd" target="_blank">a27762b</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-25
</td>
<td>
added qqplots
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/16d92267a0f0919ca70ece1ad5a7fc524c912a0a/docs/twostep_clustering_TCells_data.html" target="_blank">16d9226</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/717bb32112990968e6c864d2739f071a5aa7d2d8/analysis/twostep_clustering_TCells_data.Rmd" target="_blank">717bb32</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-12
</td>
<td>
added more descriptions
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/d51d6c4aca6f2550895732fb872de2b5a7627814/docs/twostep_clustering_TCells_data.html" target="_blank">d51d6c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/9c0697696961f66bdcbb5c01d603c1dbe454d202/analysis/twostep_clustering_TCells_data.Rmd" target="_blank">9c06976</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-12
</td>
<td>
added more descriptions
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/kevinlkx/GSFA-analysis/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/twostep_clustering_TCells_data.html" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ce094cdc91cf4d438c8a79430b8d94fd78dc0a4b/analysis/twostep_clustering_TCells_data.Rmd" target="_blank">ce094cd</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
<td>
update intro formating and barplot for DEGs
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>slurm setting</p>
<pre class="bash"><code>sinteractive --partition=broadwl --account=pi-xinhe --mem=50G --time=10:00:00 --cpus-per-task=8</code></pre>
<pre class="bash"><code>mkdir -p /project2/xinhe/kevinluo/GSFA/data

cp /project2/xinhe/yifan/Factor_analysis/shared_data/TCells_cropseq_data_seurat.rds \
  /project2/xinhe/kevinluo/GSFA/data

cp /project2/xinhe/yifan/Factor_analysis/Stimulated_T_Cells/GSE119450_RAW/D1N/genes.tsv \
  /project2/xinhe/kevinluo/GSFA/data/Stimulated_T_Cells_GSE119450_RAW_D1N_genes.tsv.gz</code></pre>
<div id="load-the-data-sets" class="section level2">
<h2>Load the data sets</h2>
<p>CROP-seq datasets: <code>/project2/xinhe/yifan/Factor_analysis/shared_data/LUHMES_cropseq_data_seurat.rds</code> The data are Seurat objects, with raw gene counts stored in <code>obj@assays$RNA@counts</code>, and cell meta data stored in <code>obj@meta.data</code>. Normalized and scaled data used for GSFA are stored in <code>obj@assays$RNA@scale.data</code>, the rownames of which are the 6k genes used for GSFA.</p>
<p>Load packages</p>
<pre class="r"><code>suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(ggplot2))
require(reshape2)
require(dplyr)
require(ComplexHeatmap)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 13),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)

library(gridExtra)
source(&quot;code/plotting_functions.R&quot;)</code></pre>
<p>Set directories</p>
<pre class="r"><code>data_dir &lt;- &quot;/project2/xinhe/kevinluo/GSFA/data/&quot;
res_dir &lt;- &quot;/project2/xinhe/kevinluo/GSFA/twostep_clustering/Stimulated_T_Cells/stimulated&quot;
dir.create(res_dir, recursive = TRUE, showWarnings = FALSE)</code></pre>
<p>Load input data</p>
<pre class="r"><code>combined_obj &lt;- readRDS(file.path(data_dir,&quot;TCells_cropseq_data_seurat.rds&quot;))</code></pre>
<p>Extract data for stimulated cells</p>
<pre class="r"><code>metadata &lt;- combined_obj@meta.data
table(metadata$orig.ident)

combined_obj@meta.data$condition &lt;- &quot;unstimulated&quot;
combined_obj@meta.data$condition[which(endsWith(combined_obj@meta.data$orig.ident, &quot;S&quot;))] &lt;- &quot;stimulated&quot;

# combined_obj.list &lt;- SplitObject(combined_obj, split.by = &quot;condition&quot;)
# combined_obj &lt;- combined_obj.list$stimulated
combined_obj &lt;- subset(combined_obj, subset = condition == &quot;stimulated&quot;)
combined_obj
table(combined_obj@meta.data$orig.ident)</code></pre>
<pre><code>
TCells_D1N TCells_D1S TCells_D2N TCells_D2S 
      5533       6843       5144       7435 
An object of class Seurat 
33694 features across 14278 samples within 1 assay 
Active assay: RNA (33694 features, 1000 variable features)
 2 dimensional reductions calculated: pca, umap

TCells_D1S TCells_D2S 
      6843       7435 </code></pre>
</div>
<div id="pre-processing" class="section level2">
<h2>Pre-processing</h2>
<p>The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat.</p>
<p>These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.</p>
<pre class="r"><code># The number of unique genes detected in each cell.
range(combined_obj$nFeature_RNA)

# The total number of molecules detected within a cell
range(combined_obj$nCount_RNA)

# The percentage of reads that map to the mitochondrial genome
range(combined_obj$percent_mt)</code></pre>
<pre><code>[1]  973 5889
[1]  2602 39985
[1] 0.01762736 9.96515679</code></pre>
<pre class="r"><code># Visualize QC metrics as a violin plot
VlnPlot(combined_obj, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent_mt&quot;), ncol = 3)</code></pre>
<p>We filter cells that have more than 500 genes identified, as in the Shifrut et al. paper.</p>
<pre class="r"><code>combined_obj &lt;- subset(combined_obj, subset = nFeature_RNA &gt; 500)</code></pre>
<p>Normalizing the data</p>
<pre class="r"><code>combined_obj &lt;- NormalizeData(combined_obj, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)</code></pre>
<p>Identification of highly variable features (feature selection)</p>
<p>Select a subset of features that exhibit high cell-to-cell variation in the dataset, by modeling the mean-variance relationship inherent in single-cell data.</p>
<p>Select the 1,000 most variable genes across cells, as in the Shifrut et al. paper.</p>
<pre class="r"><code>combined_obj &lt;- FindVariableFeatures(combined_obj, selection.method = &quot;vst&quot;, nfeatures = 1000)</code></pre>
<!-- Select the 6k genes used for GSFA in this analysis -->
<!-- ```{r select-genes} -->
<!-- # Normalized and scaled data used for GSFA, the rownames of which are the 6k genes used for GSFA -->
<!-- scaled_gene_matrix_in_gsfa <- combined_obj@assays$RNA@scale.data -->
<!-- dim(scaled_gene_matrix_in_gsfa) -->
<!-- selected_gene_id <- rownames(scaled_gene_matrix_in_gsfa) -->
<!-- all(selected_gene_id %in% rownames(combined_obj@assays$RNA@counts)) -->
<!-- ``` -->
<p>Regress out total UMI counts per cell and percent of mitochondrial genes detected per cell and scaled to obtain gene level z-scores, as in the Shifrut et al. paper.</p>
<pre class="r"><code>combined_obj &lt;- ScaleData(combined_obj, vars.to.regress = c(&quot;nCount_RNA&quot;, &quot;percent_mt&quot;))
# combined_obj &lt;- ScaleData(combined_obj, vars.to.regress = c(&quot;nCount_RNA&quot;, &quot;percent_mt&quot;), features = selected_gene_id)
dim(combined_obj[[&quot;RNA&quot;]]@counts)
dim(combined_obj[[&quot;RNA&quot;]]@data)
dim(combined_obj[[&quot;RNA&quot;]]@scale.data)

saveRDS(combined_obj, file = file.path(res_dir, &quot;TCells_stimulated_seurat_processed_data.rds&quot;))</code></pre>
</div>
<div id="perform-dimensional-reduction" class="section level2">
<h2>Perform dimensional reduction</h2>
<p>Perform PCA on the scaled data.</p>
<pre class="r"><code>combined_obj &lt;- readRDS(file.path(res_dir, &quot;TCells_stimulated_seurat_processed_data.rds&quot;))
combined_obj &lt;- RunPCA(combined_obj, features = VariableFeatures(object = combined_obj))
ElbowPlot(combined_obj, ndims = 50)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/pca-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pca-1">
Past versions of pca-1.png
</button>
</p>
<div id="fig-pca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/pca-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="cluster-the-cells" class="section level2">
<h2>Cluster the cells</h2>
<p>Embed cells in K-nearest neighbor (KNN) graph using <code>FindNeighbors()</code> using the first 30 PCs, as in the Shifrut et al. paper Then apply the Louvain algorithm to find clusters using <code>FindClusters() function</code> with default resolution (0.8).</p>
<pre class="r"><code>combined_obj &lt;- FindNeighbors(combined_obj, dims = 1:30)
combined_obj &lt;- FindClusters(combined_obj)

saveRDS(combined_obj, file = file.path(res_dir, &quot;TCells_stimulated_seurat_clustered.rds&quot;))</code></pre>
<p>Visualize the clusters using UMAP</p>
<pre class="r"><code>combined_obj &lt;- readRDS(file.path(res_dir, &quot;TCells_stimulated_seurat_clustered.rds&quot;))

cluster_labels &lt;- Idents(combined_obj)
cluster_labels &lt;- as.factor(as.numeric(as.character(cluster_labels))+1)
new_cluster_labels &lt;- paste0(&quot;k&quot;, levels(cluster_labels))
names(new_cluster_labels) &lt;- levels(combined_obj)
combined_obj &lt;- RenameIdents(combined_obj, new_cluster_labels)

combined_obj &lt;- RunUMAP(combined_obj, dims = 1:30)
DimPlot(combined_obj, reduction = &quot;umap&quot;, label = TRUE)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/umap-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-1">
Past versions of umap-1.png
</button>
</p>
<div id="fig-umap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/umap-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="finding-differentially-expressed-genes-using-mast" class="section level2">
<h2>Finding differentially expressed genes using MAST</h2>
<pre class="r"><code>combined_obj &lt;- readRDS(file.path(res_dir, &quot;TCells_stimulated_seurat_clustered.rds&quot;))

cat(&quot;Run DE test using MAST...\n&quot;)
cat(length(levels(combined_obj)), &quot;clusters.\n&quot;)
registerDoParallel(cores=n_cores)
ptm &lt;- proc.time()
de.markers &lt;- foreach(i=levels(combined_obj), .packages=&quot;Seurat&quot;) %dopar% {
  FindMarkers(combined_obj, ident.1 = i, test.use = &quot;MAST&quot;)
}
proc.time() - ptm
stopImplicitCluster()
saveRDS(de.markers, file = file.path(res_dir, &quot;TCells_stimulated_seurat_MAST_DEGs.rds&quot;))</code></pre>
</div>
<div id="associate-perturbations-with-clusters" class="section level2">
<h2>Associate perturbations with clusters</h2>
<pre class="r"><code>combined_obj &lt;- readRDS(file.path(res_dir, &quot;TCells_stimulated_seurat_clustered.rds&quot;))
perturb_matrix &lt;- combined_obj@meta.data[, 4:24]

cluster_labels &lt;- Idents(combined_obj)
cluster_labels &lt;- as.factor(as.numeric(as.character(cluster_labels))+1)
new_cluster_labels &lt;- paste0(&quot;k&quot;, levels(cluster_labels))
names(new_cluster_labels) &lt;- levels(combined_obj)
combined_obj &lt;- RenameIdents(combined_obj, new_cluster_labels)

cluster_matrix &lt;- matrix(0, nrow = nrow(perturb_matrix), ncol = length(levels(cluster_labels)))
cluster_matrix[cbind(1:nrow(perturb_matrix), cluster_labels)] &lt;- 1
rownames(cluster_matrix) &lt;- rownames(perturb_matrix)
colnames(cluster_matrix) &lt;- new_cluster_labels</code></pre>
<p>Use Chi-squared tests for the association of perturbations and clusters (2 x 2 tables)</p>
<pre class="r"><code>summary_df &lt;- expand.grid(colnames(perturb_matrix), colnames(cluster_matrix))
colnames(summary_df) &lt;- c(&quot;perturb&quot;, &quot;cluster&quot;)

summary_df &lt;- cbind(summary_df, statistic = NA, stdres = NA, pval = NA)

for(i in 1:nrow(summary_df)){
  dt &lt;- table(data.frame(perturb = perturb_matrix[,summary_df$perturb[i]], 
                         cluster = cluster_matrix[,summary_df$cluster[i]]))
  chisq &lt;- chisq.test(dt)
  summary_df[i, ]$statistic &lt;- chisq$statistic
  summary_df[i, ]$stdres &lt;- chisq$stdres[2,2]
  summary_df[i, ]$pval &lt;- chisq$p.value
}

summary_df$fdr &lt;- p.adjust(summary_df$pval, method = &quot;BH&quot;)
summary_df$bonferroni_adj &lt;- p.adjust(summary_df$pval, method = &quot;bonferroni&quot;)
saveRDS(summary_df, file = file.path(res_dir, &quot;TCells_stimulated_seurat_guide_cluster_chisq_summary_df.rds&quot;))

stdres_mat &lt;- reshape2::dcast(summary_df %&gt;% dplyr::select(perturb, cluster, stdres), perturb ~ cluster, value.var = &quot;stdres&quot;)
rownames(stdres_mat) &lt;- stdres_mat$perturb
stdres_mat$perturb &lt;- NULL

fdr_mat &lt;- reshape2::dcast(summary_df %&gt;% dplyr::select(perturb, cluster, fdr), perturb ~ cluster, value.var = &quot;fdr&quot;)
rownames(fdr_mat) &lt;- fdr_mat$perturb
fdr_mat$perturb &lt;- NULL

bonferroni_mat &lt;- reshape2::dcast(summary_df %&gt;% dplyr::select(perturb, cluster, bonferroni_adj), 
                                  perturb ~ cluster, value.var = &quot;bonferroni_adj&quot;)
rownames(bonferroni_mat) &lt;- bonferroni_mat$perturb
bonferroni_mat$perturb &lt;- NULL</code></pre>
<p>Plot perturbation ~ cluster associations (show FDR)</p>
<pre class="r"><code>KO_names &lt;- rownames(fdr_mat)
dotplot_effectsize(t(stdres_mat), t(fdr_mat),
                   reorder_markers = c(KO_names[KO_names!=&quot;NonTarget&quot;], &quot;NonTarget&quot;),
                   color_lgd_title = &quot;Chi-squared test\nstandardized residuals&quot;,
                   size_lgd_title = &quot;FDR&quot;,
                   max_score = 4,
                   min_score = -4,
                   by_score = 2) + coord_flip()</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/stdres-fdr-dotplot-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-stdres-fdr-dotplot-1">
Past versions of stdres-fdr-dotplot-1.png
</button>
</p>
<div id="fig-stdres-fdr-dotplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/stdres-fdr-dotplot-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot perturbation ~ cluster associations (show Bonferroni adjusted p-values)</p>
<pre class="r"><code>KO_names &lt;- rownames(bonferroni_mat)
dotplot_effectsize(t(stdres_mat), t(bonferroni_mat),
                   reorder_markers = c(KO_names[KO_names!=&quot;NonTarget&quot;], &quot;NonTarget&quot;),
                   color_lgd_title = &quot;Chi-squared test\nstandardized residuals&quot;,
                   size_lgd_title = &quot;Bonferroni\nadjusted p-value&quot;,
                   max_score = 4,
                   min_score = -4,
                   by_score = 2) + coord_flip()</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/stdres-bonferroni-dotplot-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-stdres-bonferroni-dotplot-1">
Past versions of stdres-bonferroni-dotplot-1.png
</button>
</p>
<div id="fig-stdres-bonferroni-dotplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/stdres-bonferroni-dotplot-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="find-de-genes-for-each-cluster-and-assign-de-genes-to-associated-perturbations" class="section level2">
<h2>Find DE genes for each cluster and assign DE genes to associated perturbations</h2>
<p>First, find DE genes for each cluster using MAST (Bonferroni adjusted p-values &lt; 0.05), Then, for each perturbation, find the associated clusters, and pull the DE genes for those clusters.</p>
<pre class="r"><code>feature.names &lt;- data.frame(fread(file.path(data_dir, &quot;Stimulated_T_Cells_GSE119450_RAW_D1N_genes.tsv.gz&quot;),
                                  header = FALSE), stringsAsFactors = FALSE)

de.markers &lt;- readRDS(file.path(res_dir, &quot;TCells_stimulated_seurat_MAST_DEGs.rds&quot;))
names(de.markers) &lt;- paste0(&quot;k&quot;, levels(cluster_labels))

de.genes.clusters &lt;- vector(&quot;list&quot;, length = length(de.markers))
names(de.genes.clusters) &lt;- names(de.markers)
for( i in 1:length(de.genes.clusters)){
  de_sumstats &lt;- de.markers[[i]]
  de_genes &lt;- unique(rownames(de_sumstats[de_sumstats$p_val_adj &lt; 0.05,]))
  # de_genes &lt;- feature.names$V2[match(de_genes, feature.names$V1)]
  de.genes.clusters[[i]] &lt;- de_genes
}</code></pre>
<p>Number of DE genes for each perturbation (Chi-squared test FDR &lt; 0.05)</p>
<pre class="r"><code>perturb_names &lt;- rownames(fdr_mat)
perturb_names &lt;- c(&quot;NonTarget&quot;, perturb_names[perturb_names!=&quot;NonTarget&quot;])

de.genes.perturbs &lt;- vector(&quot;list&quot;, length = length(perturb_names))
names(de.genes.perturbs) &lt;- perturb_names

for(i in 1:length(de.genes.perturbs)){
  perturb_name &lt;- names(de.genes.perturbs)[i]
  associated_cluster_labels &lt;- colnames(fdr_mat)[which(fdr_mat[perturb_name, ] &lt; 0.05)]
  if(length(associated_cluster_labels) &gt; 0){
    de.genes.perturbs[[i]] &lt;- unique(unlist(de.genes.clusters[associated_cluster_labels]))
  }
}

num.de.genes.perturbs &lt;- sapply(de.genes.perturbs, length)
twostep_clustering_fdr0.05_genes &lt;- de.genes.perturbs

dge_plot_df &lt;- data.frame(Perturbation = names(num.de.genes.perturbs), Num_DEGs = num.de.genes.perturbs)
dge_plot_df$Perturbation &lt;- factor(dge_plot_df$Perturbation, levels = names(num.de.genes.perturbs))

ggplot(data=dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1)) +
  geom_bar(position = &quot;dodge&quot;, stat = &quot;identity&quot;) +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = &quot;Set2&quot;) +
  labs(x = &quot;Target gene&quot;,
       y = &quot;Number of DEGs&quot;,
       title = &quot;Number of DEGs detected by Two-step clustering with MAST DE analysis&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = &quot;bottom&quot;,
        legend.text = element_text(size = 13))</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/count-de-genes-fdr-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-count-de-genes-fdr-1">
Past versions of count-de-genes-fdr-1.png
</button>
</p>
<div id="fig-count-de-genes-fdr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/count-de-genes-fdr-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Number of DE genes for each perturbation (Chi-squared test Bonferroni adjusted p-value &lt; 0.05)</p>
<pre class="r"><code>perturb_names &lt;- rownames(bonferroni_mat)
perturb_names &lt;- c(&quot;NonTarget&quot;, perturb_names[perturb_names!=&quot;NonTarget&quot;])

de.genes.perturbs &lt;- vector(&quot;list&quot;, length = length(perturb_names))
names(de.genes.perturbs) &lt;- perturb_names

for(i in 1:length(de.genes.perturbs)){
  perturb_name &lt;- names(de.genes.perturbs)[i]
  associated_cluster_labels &lt;- colnames(bonferroni_mat)[which(bonferroni_mat[perturb_name, ] &lt; 0.05)]
  if(length(associated_cluster_labels) &gt; 0){
    de.genes.perturbs[[i]] &lt;- unique(unlist(de.genes.clusters[associated_cluster_labels]))
  }
}

num.de.genes.perturbs &lt;- sapply(de.genes.perturbs, length)
twostep_clustering_bonferroni0.05_genes &lt;- de.genes.perturbs

dge_plot_df &lt;- data.frame(Perturbation = names(num.de.genes.perturbs), Num_DEGs = num.de.genes.perturbs)
dge_plot_df$Perturbation &lt;- factor(dge_plot_df$Perturbation, levels = names(num.de.genes.perturbs))

ggplot(data=dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1)) +
  geom_bar(position = &quot;dodge&quot;, stat = &quot;identity&quot;) +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = &quot;Set2&quot;) +
  labs(x = &quot;Target gene&quot;,
       y = &quot;Number of DEGs&quot;,
       title = &quot;Number of DEGs detected by Two-step clustering with MAST DE analysis&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = &quot;bottom&quot;,
        legend.text = element_text(size = 13))</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/count-de-genes-bonferroni-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-count-de-genes-bonferroni-1">
Past versions of count-de-genes-bonferroni-1.png
</button>
</p>
<div id="fig-count-de-genes-bonferroni-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/ee884c45858acc41f9e7b9c4ab878029576c1a51/docs/figure/twostep_clustering_TCells_data.Rmd/count-de-genes-bonferroni-1.png" target="_blank">ee884c4</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="compare-single-gene-de-p-value-distributions-between-two-step-clustering-analysis-and-gsfa" class="section level2">
<h2>Compare single-gene DE p-value distributions between two-step clustering analysis and GSFA</h2>
<pre class="r"><code>fdr_cutoff &lt;- 0.05
lfsr_cutoff &lt;- 0.05</code></pre>
<p>Load the output of GSFA <code>fit_gsfa_multivar()</code> run.</p>
<pre class="r"><code>data_folder &lt;- &quot;/project2/xinhe/yifan/Factor_analysis/Stimulated_T_Cells/&quot;
fit &lt;- readRDS(paste0(data_folder,
                      &quot;gsfa_output_detect_01/all_uncorrected_by_group.use_negctrl/All.gibbs_obj_k20.svd_negctrl.restart.light.rds&quot;))
gibbs_PM &lt;- fit$posterior_means
lfsr_mat1 &lt;- fit$lfsr1[, -ncol(fit$lfsr1)]
lfsr_mat0 &lt;- fit$lfsr0[, -ncol(fit$lfsr0)]
total_effect1 &lt;- fit$total_effect1[, -ncol(fit$total_effect1)]
total_effect0 &lt;- fit$total_effect0[, -ncol(fit$total_effect0)]
KO_names &lt;- colnames(lfsr_mat1)</code></pre>
<p>DEGs detected by GSFA</p>
<pre><code>   ARID1A      BTLA  C10orf54      CBLB      CD3D       CD5    CDKN1B      DGKA 
      393       107        66       631         0       645       468        32 
     DGKZ    HAVCR2      LAG3      LCP2     MEF2D NonTarget     PDCD1     RASA2 
      113        35         1       589        15         0         0       277 
    SOCS1     STAT6     TCEB2   TMEM222   TNFRSF9 
      356         1       300         4        14 </code></pre>
<p>Load MAST single-gene DE result</p>
<pre class="r"><code>guides &lt;- KO_names[KO_names!=&quot;NonTarget&quot;]

mast_list &lt;- list()
for (m in guides){
  fname &lt;- paste0(data_folder, &quot;processed_data/MAST/all_by_stim_negctrl/gRNA_&quot;, 
                  m, &quot;.dev_res_top6k.vs_negctrl.rds&quot;)
  tmp_df &lt;- readRDS(fname)
  tmp_df$geneID &lt;- rownames(tmp_df)
  tmp_df &lt;- tmp_df %&gt;% dplyr::rename(FDR = fdr, PValue = pval)
  mast_list[[m]] &lt;- tmp_df
}
mast_signif_counts &lt;- sapply(mast_list, function(x){filter(x, FDR &lt; fdr_cutoff) %&gt;% nrow()})</code></pre>
<p>QQ-plots of MAST DE p-values for the GSFA significant genes vs two-step clustering DE genes.</p>
<pre class="r"><code>qqplots &lt;- list()
for(i in 1:length(guides)){
  guide &lt;- guides[i]
  mast_res &lt;- mast_list[[guide]]
  tsc_de_genes &lt;- twostep_clustering_fdr0.05_genes[[guide]]
  gsfa_de_genes &lt;- gsfa_sig_genes[[guide]]
  tsc_de_genes &lt;- intersect(tsc_de_genes, rownames(mast_res))
  gsfa_de_genes &lt;- intersect(gsfa_de_genes, rownames(mast_res))
  
  if(length(tsc_de_genes)&gt;0 &amp;&amp; length(gsfa_de_genes) &gt;0){
    cat(&quot;plot&quot;, guide, &quot;\n&quot;)
    mast_res$tsc_gene &lt;- 0
    mast_res[tsc_de_genes, ]$tsc_gene &lt;- 1
    mast_res$gsfa_gene &lt;- 0
    mast_res[gsfa_de_genes, ]$gsfa_gene &lt;- 1
    pvalue_list &lt;- list(&#39;Two-step clustering&#39;=dplyr::filter(mast_res,tsc_gene==1)$PValue,
                        &#39;GSFA&#39;=dplyr::filter(mast_res,gsfa_gene==1)$PValue,
                        &#39;MAST&#39;=mast_res$PValue)
    qqplots[[guide]] &lt;- qqplot.pvalue(pvalue_list, pointSize = 1, legendSize = 4) + 
      ggtitle(guide) + theme(plot.title = element_text(hjust = 0.5)) +
      scale_colour_discrete(name=&quot;Method&quot;)
  }
}

grid.arrange(grobs = qqplots, nrow = 3, ncol = 2)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-all-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-qqplots-gsfa-tsc-all-1">
Past versions of qqplots-gsfa-tsc-all-1.png
</button>
</p>
<div id="fig-qqplots-gsfa-tsc-all-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/104e3915d02deea37cfb2afda409088498058f32/docs/figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-all-1.png" target="_blank">104e391</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-25
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre><code>plot CBLB 
plot CD5 
plot CDKN1B 
plot LCP2 
plot TCEB2 
plot TNFRSF9 </code></pre>
<p>Pooling p-values from all perturbations</p>
<pre class="r"><code>combined_mast_res &lt;- data.frame()
for(i in 1:length(guides)){
  guide &lt;- guides[i]
  mast_res &lt;- mast_list[[guide]]
  tsc_de_genes &lt;- twostep_clustering_fdr0.05_genes[[guide]]
  gsfa_de_genes &lt;- gsfa_sig_genes[[guide]]
  tsc_de_genes &lt;- intersect(tsc_de_genes, rownames(mast_res))
  gsfa_de_genes &lt;- intersect(gsfa_de_genes, rownames(mast_res))
  mast_res$tsc_gene &lt;- 0
  if(length(tsc_de_genes) &gt;0){
    mast_res[tsc_de_genes, ]$tsc_gene &lt;- 1
  }
  mast_res$gsfa_gene &lt;- 0
  if(length(gsfa_de_genes) &gt;0){
    mast_res[gsfa_de_genes, ]$gsfa_gene &lt;- 1
  }
  combined_mast_res &lt;- rbind(combined_mast_res, mast_res)
}

pvalue_list &lt;- list(&#39;Two-step clustering&#39;=dplyr::filter(combined_mast_res,tsc_gene==1)$PValue,
                    &#39;GSFA&#39;=dplyr::filter(combined_mast_res,gsfa_gene==1)$PValue,
                    &#39;MAST&#39;=combined_mast_res$PValue)

# pdf(file.path(res_dir, &quot;qqplot_all_combined.pdf&quot;))
qqplot.pvalue(pvalue_list, pointSize = 1, legendSize = 4) + 
      ggtitle(&quot;All perturbations&quot;) + theme(plot.title = element_text(hjust = 0.5)) +
      scale_colour_discrete(name=&quot;Method&quot;)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-all-combined-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># dev.off()</code></pre>
<p>QQ-plots of MAST DE p-values for the GSFA only genes vs two-step only genes.</p>
<pre class="r"><code>qqplots &lt;- list()
for(i in 1:length(guides)){
  guide &lt;- guides[i]
  mast_res &lt;- mast_list[[guide]]
  tsc_de_genes &lt;- twostep_clustering_fdr0.05_genes[[guide]]
  gsfa_de_genes &lt;- gsfa_sig_genes[[guide]]
  tsc_de_genes &lt;- intersect(tsc_de_genes, rownames(mast_res))
  gsfa_de_genes &lt;- intersect(gsfa_de_genes, rownames(mast_res))
  
  if(length(tsc_de_genes)&gt;0 &amp;&amp; length(gsfa_de_genes) &gt;0){
    cat(&quot;plot&quot;, guide, &quot;\n&quot;)
    mast_res$tsc_only_gene &lt;- 0
    mast_res[setdiff(tsc_de_genes, gsfa_de_genes), ]$tsc_only_gene &lt;- 1
    mast_res$gsfa_only_gene &lt;- 0
    mast_res[setdiff(gsfa_de_genes, tsc_de_genes), ]$gsfa_only_gene &lt;- 1
    
    pvalue_list &lt;- list(&#39;Two-step clustering only&#39;=dplyr::filter(mast_res,tsc_only_gene==1)$PValue,
                        &#39;GSFA only&#39;=dplyr::filter(mast_res,gsfa_only_gene==1)$PValue,
                        &#39;MAST&#39;=mast_res$PValue)
    qqplots[[guide]] &lt;- qqplot.pvalue(pvalue_list, pointSize = 1, legendSize = 4) + 
      ggtitle(guide) + theme(plot.title = element_text(hjust = 0.5)) +
      scale_colour_discrete(name=&quot;Method&quot;)
  }
}

grid.arrange(grobs = qqplots, nrow = 3, ncol = 2)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-only-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-qqplots-gsfa-tsc-only-1">
Past versions of qqplots-gsfa-tsc-only-1.png
</button>
</p>
<div id="fig-qqplots-gsfa-tsc-only-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/kevinlkx/GSFA-analysis/blob/104e3915d02deea37cfb2afda409088498058f32/docs/figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-only-1.png" target="_blank">104e391</a>
</td>
<td>
kevinlkx
</td>
<td>
2022-08-25
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre><code>plot CBLB 
plot CD5 
plot CDKN1B 
plot LCP2 
plot TCEB2 
plot TNFRSF9 </code></pre>
<p>Pooling p-values from all perturbations</p>
<pre class="r"><code>combined_mast_res &lt;- data.frame()
for(i in 1:length(guides)){
  guide &lt;- guides[i]
  mast_res &lt;- mast_list[[guide]]
  tsc_de_genes &lt;- twostep_clustering_fdr0.05_genes[[guide]]
  gsfa_de_genes &lt;- gsfa_sig_genes[[guide]]
  tsc_de_genes &lt;- intersect(tsc_de_genes, rownames(mast_res))
  gsfa_de_genes &lt;- intersect(gsfa_de_genes, rownames(mast_res))
  mast_res$tsc_only_gene &lt;- 0
  if(length(setdiff(tsc_de_genes, gsfa_de_genes)) &gt;0){
    mast_res[setdiff(tsc_de_genes, gsfa_de_genes), ]$tsc_only_gene &lt;- 1
  }
  mast_res$gsfa_only_gene &lt;- 0
  if(length(setdiff(gsfa_de_genes, tsc_de_genes)) &gt;0){
    mast_res[setdiff(gsfa_de_genes, tsc_de_genes), ]$gsfa_only_gene &lt;- 1
  }
  combined_mast_res &lt;- rbind(combined_mast_res, mast_res)
}

pvalue_list &lt;- list(&#39;Two-step clustering only&#39;=dplyr::filter(combined_mast_res,tsc_only_gene==1)$PValue,
                    &#39;GSFA only&#39;=dplyr::filter(combined_mast_res,gsfa_only_gene==1)$PValue,
                    &#39;MAST&#39;=combined_mast_res$PValue)

# pdf(file.path(res_dir, &quot;qqplot_only_combined.pdf&quot;))
qqplot.pvalue(pvalue_list, pointSize = 1, legendSize = 4) + 
      ggtitle(&quot;All perturbations&quot;) + theme(plot.title = element_text(hjust = 0.5)) +
      scale_colour_discrete(name=&quot;Method&quot;)</code></pre>
<p><img src="figure/twostep_clustering_TCells_data.Rmd/qqplots-gsfa-tsc-only-combined-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># dev.off()</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.4 (2021-02-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] lattice_0.20-45      gridExtra_2.3        dplyr_1.0.8         
 [4] reshape2_1.4.4       ggplot2_3.3.5        ComplexHeatmap_2.6.2
 [7] SeuratObject_4.0.4   Seurat_4.1.0         data.table_1.14.2   
[10] workflowr_1.7.0     

loaded via a namespace (and not attached):
  [1] circlize_0.4.15       plyr_1.8.6            igraph_1.3.4         
  [4] lazyeval_0.2.2        splines_4.0.4         listenv_0.8.0        
  [7] scattermore_0.7       digest_0.6.29         htmltools_0.5.3      
 [10] fansi_1.0.3           magrittr_2.0.3        tensor_1.5           
 [13] cluster_2.1.3         ROCR_1.0-11           globals_0.16.0       
 [16] matrixStats_0.62.0    R.utils_2.12.0        spatstat.sparse_2.1-0
 [19] colorspace_2.0-3      ggrepel_0.9.1         xfun_0.30            
 [22] callr_3.7.0           crayon_1.5.1          jsonlite_1.8.0       
 [25] spatstat.data_2.1-2   survival_3.3-1        zoo_1.8-9            
 [28] glue_1.6.2            polyclip_1.10-0       gtable_0.3.0         
 [31] leiden_0.3.9          GetoptLong_1.0.5      future.apply_1.8.1   
 [34] shape_1.4.6           BiocGenerics_0.36.1   abind_1.4-5          
 [37] scales_1.2.0          DBI_1.1.3             spatstat.random_2.1-0
 [40] miniUI_0.1.1.1        Rcpp_1.0.9            viridisLite_0.4.0    
 [43] xtable_1.8-4          clue_0.3-60           reticulate_1.25      
 [46] spatstat.core_2.4-0   stats4_4.0.4          htmlwidgets_1.5.4    
 [49] httr_1.4.2            RColorBrewer_1.1-3    ellipsis_0.3.2       
 [52] ica_1.0-2             R.methodsS3_1.8.1     pkgconfig_2.0.3      
 [55] farver_2.1.1          sass_0.4.1            uwot_0.1.11          
 [58] deldir_1.0-6          utf8_1.2.2            tidyselect_1.1.2     
 [61] labeling_0.4.2        rlang_1.0.4           later_1.3.0          
 [64] munsell_0.5.0         tools_4.0.4           cli_3.3.0            
 [67] generics_0.1.3        ggridges_0.5.3        evaluate_0.16        
 [70] stringr_1.4.0         fastmap_1.1.0         yaml_2.3.5           
 [73] goftest_1.2-3         processx_3.5.3        knitr_1.38           
 [76] fs_1.5.2              fitdistrplus_1.1-8    purrr_0.3.4          
 [79] RANN_2.6.1            pbapply_1.5-0         future_1.24.0        
 [82] nlme_3.1-159          whisker_0.4           mime_0.12            
 [85] R.oo_1.24.0           compiler_4.0.4        rstudioapi_0.13      
 [88] plotly_4.10.0         png_0.1-7             spatstat.utils_2.3-0 
 [91] tibble_3.1.6          bslib_0.3.1           stringi_1.7.6        
 [94] highr_0.9             ps_1.7.1              RSpectra_0.16-0      
 [97] Matrix_1.4-1          vctrs_0.4.1           pillar_1.8.0         
[100] lifecycle_1.0.1       spatstat.geom_2.3-2   lmtest_0.9-40        
[103] jquerylib_0.1.4       GlobalOptions_0.1.2   RcppAnnoy_0.0.19     
[106] cowplot_1.1.1         irlba_2.3.5           httpuv_1.6.5         
[109] patchwork_1.1.1       R6_2.5.1              promises_1.2.0.1     
[112] KernSmooth_2.23-20    IRanges_2.24.1        parallelly_1.32.1    
[115] codetools_0.2-18      MASS_7.3-58.1         assertthat_0.2.1     
[118] rprojroot_2.0.2       rjson_0.2.21          withr_2.5.0          
[121] sctransform_0.3.3     S4Vectors_0.28.1      mgcv_1.8-39          
[124] parallel_4.0.4        rpart_4.1-15          tidyr_1.2.0          
[127] rmarkdown_2.13        Cairo_1.6-0           Rtsne_0.15           
[130] git2r_0.30.1          getPass_0.2-2         shiny_1.7.1          </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
