---
title: "Two step clustering analysis on LUHMES CROP-seq data"
author: "Kaixuan Luo"
date: "2022-08-10"
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
editor_options:
  chunk_output_type: console
---

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(results = "hold", fig.align = "center", message=FALSE, warning=FALSE)
```

```{bash, eval=FALSE}
mkdir -p /project2/xinhe/kevinluo/GSFA/data

cp /project2/xinhe/yifan/Factor_analysis/shared_data/LUHMES_cropseq_data_seurat.rds \
  /project2/xinhe/kevinluo/GSFA/data

cp /project2/xinhe/yifan/Factor_analysis/LUHMES/GSE142078_raw/GSM4219575_Run1_genes.tsv.gz \
  /project2/xinhe/kevinluo/GSFA/data/LUHMES_GSM4219575_Run1_genes.tsv.gz
```

## Load the data sets

CROP-seq datasets: `/project2/xinhe/yifan/Factor_analysis/shared_data/LUHMES_cropseq_data_seurat.rds`

The data are Seurat objects, with raw gene counts stored in `obj@assays$RNA@counts`, 
and cell meta data stored in `obj@meta.data`.
Normalized and scaled data used for GSFA are stored in `obj@assays$RNA@scale.data`, 
the rownames of which are the 6k genes used for GSFA.

Load packages
```{r load-packages, message=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(MUSIC))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(ggplot2))
require(reshape2)
require(dplyr)
require(ComplexHeatmap)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 13),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)

source("code/plotting_functions.R")
```

Set directories
```{r dir}
data_dir <- "/project2/xinhe/kevinluo/GSFA/data/"
res_dir <- "/project2/xinhe/kevinluo/GSFA/twostep_clustering/LUHMES"
dir.create(res_dir, recursive = TRUE, showWarnings = FALSE)
```

Load input data
```{r load-data}
combined_obj <- readRDS(file.path(data_dir,"LUHMES_cropseq_data_seurat.rds"))
```

## pre-processing
The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat.

These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, 
and the detection of highly variable features.

QC and selecting cells for further analysis.

```{r qc-1}
# The number of unique genes detected in each cell.
cat("The number of unique genes detected in each cell:\n")
range(combined_obj$nFeature_RNA)

# The total number of molecules detected within a cell
cat("The total number of molecules detected within a cell:\n")
range(combined_obj$nCount_RNA)

# The percentage of reads that map to the mitochondrial genome
cat("The percentage of reads that map to the mitochondrial genome:\n")
range(combined_obj$percent_mt)
```

```{r qc-2, eval=FALSE}
# Visualize QC metrics as a violin plot
VlnPlot(combined_obj, features = c("nFeature_RNA", "nCount_RNA", "percent_mt"), ncol = 3)
```

We filter cells that have more than 500 genes identified.
```{r qc-filter-cells, eval=FALSE}
combined_obj <- subset(combined_obj, subset = nFeature_RNA > 500)
```

### Normalizing the data

```{r normalize-data, eval=FALSE}
combined_obj <- NormalizeData(combined_obj, normalization.method = "LogNormalize", scale.factor = 10000)
```

### Identification of highly variable features (feature selection)

Select a subset of features that exhibit high cell-to-cell variation in the dataset, by modeling the mean-variance relationship inherent in single-cell data.

Select the 1,000 most variable genes across cells. 
```{r select-variable-genes, eval=FALSE}
combined_obj <- FindVariableFeatures(combined_obj, selection.method = "vst", nfeatures = 1000)
```

<!-- Select the 6k genes used for GSFA in this analysis -->
<!-- ```{r select-genes} -->
<!-- # Normalized and scaled data used for GSFA, the rownames of which are the 6k genes used for GSFA -->
<!-- scaled_gene_matrix_in_gsfa <- combined_obj@assays$RNA@scale.data -->
<!-- dim(scaled_gene_matrix_in_gsfa) -->
<!-- selected_gene_id <- rownames(scaled_gene_matrix_in_gsfa) -->
<!-- all(selected_gene_id %in% rownames(combined_obj@assays$RNA@counts)) -->
<!-- ``` -->

### Regress out total UMI counts per cell and percent of mitochondrial genes detected per cell and scaled to obtain gene level z-scores. 

The results of this are stored in combined_obj[["RNA"]]@scale.data

```{r regress-out-totalcounts-mt-percent, eval=FALSE}
combined_obj <- ScaleData(combined_obj, vars.to.regress = c("nCount_RNA", "percent_mt"))
# combined_obj <- ScaleData(combined_obj, vars.to.regress = c("nCount_RNA", "percent_mt"), features = selected_gene_id)
dim(combined_obj[["RNA"]]@counts)
dim(combined_obj[["RNA"]]@data)
dim(combined_obj[["RNA"]]@scale.data)

saveRDS(combined_obj, file = file.path(res_dir, "LUHMES_seurat_processed_data.rds"))
```

## Perform dimensional reduction

Perform PCA on the scaled data. 
```{r pca}
combined_obj <- readRDS(file.path(res_dir, "LUHMES_seurat_processed_data.rds"))
combined_obj <- RunPCA(combined_obj, features = VariableFeatures(object = combined_obj))
ElbowPlot(combined_obj, ndims = 50)
```

## Cluster the cells

```{r clustering, eval=FALSE}
combined_obj <- FindNeighbors(combined_obj, dims = 1:30)
combined_obj <- FindClusters(combined_obj)

saveRDS(combined_obj, file = file.path(res_dir, "LUHMES_seurat_clustered.rds"))
```

## Run non-linear dimensional reduction (UMAP/tSNE)
```{r umap, message=FALSE}
combined_obj <- readRDS(file.path(res_dir, "LUHMES_seurat_clustered.rds"))

cluster_labels <- Idents(combined_obj)
cluster_labels <- as.factor(as.numeric(as.character(cluster_labels))+1)
new_cluster_labels <- paste0("k", levels(cluster_labels))
names(new_cluster_labels) <- levels(combined_obj)
combined_obj <- RenameIdents(combined_obj, new_cluster_labels)

combined_obj <- RunUMAP(combined_obj, dims = 1:30)
DimPlot(combined_obj, reduction = "umap", label = TRUE)
```

## Finding differentially expressed features (cluster biomarkers)

```{r de-markers, eval=FALSE}
combined_obj <- readRDS(file.path(res_dir, "LUHMES_seurat_clustered.rds"))

cat("Run DE test using MAST...\n")
cat(length(levels(combined_obj)), "clusters.\n")
registerDoParallel(cores=n_cores)
ptm <- proc.time()
de.markers <- foreach(i=levels(combined_obj), .packages="Seurat") %dopar% {
  FindMarkers(combined_obj, ident.1 = i, test.use = "MAST")
}
proc.time() - ptm
stopImplicitCluster()
saveRDS(de.markers, file = file.path(res_dir, "LUHMES_seurat_MAST_DEGs.rds"))
```

## Associate perturbation with clustering results

```{r associate-perturbs-clusters}
combined_obj <- readRDS(file.path(res_dir, "LUHMES_seurat_clustered.rds"))
perturb_matrix <- combined_obj@meta.data[, 4:18]

cluster_labels <- Idents(combined_obj)
cluster_labels <- as.factor(as.numeric(as.character(cluster_labels))+1)
new_cluster_labels <- paste0("k", levels(cluster_labels))
names(new_cluster_labels) <- levels(combined_obj)
combined_obj <- RenameIdents(combined_obj, new_cluster_labels)

cluster_matrix <- matrix(0, nrow = nrow(perturb_matrix), ncol = length(levels(cluster_labels)))
cluster_matrix[cbind(1:nrow(perturb_matrix), cluster_labels)] <- 1
rownames(cluster_matrix) <- rownames(perturb_matrix)
colnames(cluster_matrix) <- new_cluster_labels
```

Use Chi-squared tests for the association of perturbations and clusters
```{r chisq-test}
summary_df <- expand.grid(colnames(perturb_matrix), colnames(cluster_matrix))
colnames(summary_df) <- c("perturb", "cluster")

summary_df <- cbind(summary_df, statistic = NA, stdres = NA, pval = NA)

for(i in 1:nrow(summary_df)){
  dt <- table(data.frame(perturb = perturb_matrix[,summary_df$perturb[i]], 
                         cluster = cluster_matrix[,summary_df$cluster[i]]))
  chisq <- chisq.test(dt)
  summary_df[i, ]$statistic <- chisq$statistic
  summary_df[i, ]$stdres <- chisq$stdres[2,2]
  summary_df[i, ]$pval <- chisq$p.value
}

summary_df$fdr <- p.adjust(summary_df$pval, method = "BH")
summary_df$bonferroni_adj <- p.adjust(summary_df$pval, method = "bonferroni")

stdres_mat <- reshape2::dcast(summary_df %>% dplyr::select(perturb, cluster, stdres), perturb ~ cluster, value.var = "stdres")
rownames(stdres_mat) <- stdres_mat$perturb
stdres_mat$perturb <- NULL

fdr_mat <- reshape2::dcast(summary_df %>% dplyr::select(perturb, cluster, fdr), perturb ~ cluster, value.var = "fdr")
rownames(fdr_mat) <- fdr_mat$perturb
fdr_mat$perturb <- NULL

bonferroni_mat <- reshape2::dcast(summary_df %>% dplyr::select(perturb, cluster, bonferroni_adj), 
                                  perturb ~ cluster, value.var = "bonferroni_adj")
rownames(bonferroni_mat) <- bonferroni_mat$perturb
bonferroni_mat$perturb <- NULL
```

Prturbation ~ cluster associations (show FDR)
```{r stdres-fdr-dotplot, eval=TRUE, fig.width=9, fig.height=5.5}
KO_names <- rownames(fdr_mat)
dotplot_effectsize(t(stdres_mat), t(fdr_mat),
                   reorder_markers = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"),
                   color_lgd_title = "Chi-squared test\nstandardized residuals",
                   size_lgd_title = "FDR",
                   max_score = 4,
                   min_score = -4,
                   by_score = 2) + coord_flip()
```

Prturbation ~ cluster associations (show Bonferroni adjusted p-values)
```{r stdres-bonferroni-dotplot, eval=TRUE, fig.width=9, fig.height=5.5}
KO_names <- rownames(bonferroni_mat)
dotplot_effectsize(t(stdres_mat), t(bonferroni_mat),
                   reorder_markers = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"),
                   color_lgd_title = "Chi-squared test\nstandardized residuals",
                   size_lgd_title = "Bonferroni\nadjusted p-value",
                   max_score = 4,
                   min_score = -4,
                   by_score = 2) + coord_flip()
```

## Find DE genes for each cluster and assign DE genes to associated perturbations

First, find DE genes for each cluster using MAST (Bonferroni adjusted p-values < 0.05),
Then, for each perturbation, find the associated clusters, and pull the DE genes for those clusters.

```{r get-de-genes}
feature.names <- data.frame(fread(file.path(data_dir, "LUHMES_GSM4219575_Run1_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)

de.markers <- readRDS(file.path(res_dir, "LUHMES_seurat_MAST_DEGs.rds"))
names(de.markers) <- paste0("k", levels(cluster_labels))

de.genes.clusters <- vector("list", length = length(de.markers))
names(de.genes.clusters) <- names(de.markers)
for( i in 1:length(de.genes.clusters)){
  de_sumstats <- de.markers[[i]]
  de_genes <- unique(rownames(de_sumstats[de_sumstats$p_val_adj < 0.05,]))
  de_genes <- feature.names$V2[match(de_genes, feature.names$V1)]
  de.genes.clusters[[i]] <- de_genes
}
```


Number of DE genes for each perturbation (Chi-squared test FDR < 0.05)
```{r count-de-genes-fdr, fig.width=13, fig.height=5}
perturb_names <- rownames(fdr_mat)
perturb_names <- c("Nontargeting", perturb_names[perturb_names!="Nontargeting"])

de.genes.perturbs <- vector("list", length = length(perturb_names))
names(de.genes.perturbs) <- perturb_names

for(i in 1:length(de.genes.perturbs)){
  perturb_name <- names(de.genes.perturbs)[i]
  associated_cluster_labels <- colnames(fdr_mat)[which(fdr_mat[perturb_name, ] < 0.05)]
  if(length(associated_cluster_labels) > 0){
    de.genes.perturbs[[i]] <- unique(unlist(de.genes.clusters[associated_cluster_labels]))
  }
}

num.de.genes.perturbs <- sapply(de.genes.perturbs, length)

dge_plot_df <- data.frame(Perturbation = names(num.de.genes.perturbs), Num_DEGs = num.de.genes.perturbs)
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation, levels = names(num.de.genes.perturbs))

ggplot(data=dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Target gene",
       y = "Number of DEGs",
       title = "Number of DEGs detected by Two-step clustering with MAST DE analysis") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

Number of DE genes for each perturbation (Chi-squared test Bonferroni adjusted p-value < 0.05)
```{r count-de-genes-bonferroni, fig.width=13, fig.height=5}

perturb_names <- rownames(bonferroni_mat)
perturb_names <- c("Nontargeting", perturb_names[perturb_names!="Nontargeting"])

de.genes.perturbs <- vector("list", length = length(perturb_names))
names(de.genes.perturbs) <- perturb_names

for(i in 1:length(de.genes.perturbs)){
  perturb_name <- names(de.genes.perturbs)[i]
  associated_cluster_labels <- colnames(bonferroni_mat)[which(bonferroni_mat[perturb_name, ] < 0.05)]
  if(length(associated_cluster_labels) > 0){
    de.genes.perturbs[[i]] <- unique(unlist(de.genes.clusters[associated_cluster_labels]))
  }
}

num.de.genes.perturbs <- sapply(de.genes.perturbs, length)

dge_plot_df <- data.frame(Perturbation = names(num.de.genes.perturbs), Num_DEGs = num.de.genes.perturbs)
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation, levels = names(num.de.genes.perturbs))

ggplot(data=dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Target gene",
       y = "Number of DEGs",
       title = "Number of DEGs detected by Two-step clustering with MAST DE analysis") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

